### PR TITLE
feat: add support for comments in adhoc clauses

### PR DESCRIPTION
**Pull request by villebro**
### SUMMARY
Add support for comments in adhoc clauses to make it easier to comment out sections of adhoc metrics/filters when developing charts. This is done by making sure that all freeform queries containing a single line comment `--` are suffixed with a line change.

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (follow approval process in [SIP-59](https://github.com/apache/superset/issues/13351))
  - [ ] Migration is atomic, supports rollback & is backwards-compatible
  - [ ] Confirm DB migration upgrade and downgrade tested
  - [ ] Runtime estimates and downtime expectations provided
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/sql_parse.py [hunk 1]
```diff
@@ -441,25 +442,35 @@ def set_or_update_query_limit(self, new_limit: int, force: bool = False) -> str:
...
-def validate_filter_clause(clause: str) -> None:
-    if sqlparse.format(clause, strip_comments=True) != sqlparse.format(clause):
-        raise QueryClauseValidationException("Filter clause contains comment")
-
+def sanitize_clause(clause: str) -> str:
+    # clause = sqlparse.format(clause, strip_comments=True)
     statements = sqlparse.parse(clause)
     if len(statements) != 1:
-        raise QueryClauseValidationException("Filter clause contains multiple queries")
+        raise QueryClauseValidationException("Clause contains multiple queries")
```
Comment (betodealmeida) [inline superset/sql_parse.py:449]: Nit:

```suggestion
        raise QueryClauseValidationException("Clause contains multiple statements")
```

**Review by betodealmeida — APPROVED** [review]
This looks great!

File: superset-frontend/packages/superset-ui-core/src/query/processFilters.ts [hunk 2]
```diff
@@ -23,6 +23,14 @@ import { QueryObjectFilterClause } from './types/Query';
 import { isSimpleAdhocFilter } from './types/Filter';
 import convertFilter from './convertFilter';
 
+function sanitizeClause(clause: string): string {
```
Comment (suddjian) [inline superset-frontend/packages/superset-ui-core/src/query/processFilters.ts:26]: This needs to be done on backend, no?

File: superset/sql_parse.py [hunk 3]
```diff
@@ -441,25 +442,35 @@ def set_or_update_query_limit(self, new_limit: int, force: bool = False) -> str:
...
                 raise QueryClauseValidationException(
                     "Closing unclosed parenthesis in filter clause"
                 )
+        previous_token = token
     if open_parens > 0:
         raise QueryClauseValidationException("Unclosed parenthesis in filter clause")
 
+    if previous_token and previous_token.ttype in Comment:
+        if previous_token.value[-1] != "\n":
+            clause = f"{clause}\n"
```
Comment (suddjian) [inline superset/sql_parse.py:470]: Ah, I see it is done in the backend. But I still don't understand why we also do this in the frontend?

**Review by suddjian — APPROVED** [review]

**Comment by sadpandajoe** [comment]
🏷️ preset:2022.11

**Pull request merged**